homeassistant/core.py

```diff
@@ -266,36 +245,25 @@ def _loop_empty(self):
 
     def block_till_done(self):
         """Block till all pending work is done."""
-        complete = threading.Event()
+        run_coroutine_threadsafe(
+            self.async_block_till_done(), loop=self.loop).retsult()
 
-        @asyncio.coroutine
-        def sleep_wait():
-            """Sleep in thread pool."""
-            yield from self.loop.run_in_executor(None, time.sleep, 0)
-
-        def notify_when_done():
-            """Notify event loop when pool done."""
-            count = 0
-            while True:
-                # Wait for the work queue to empty
-                self.pool.block_till_done()
-
-                # Verify the loop is empty
-                if self._loop_empty():
-                    count += 1
-
-                if count == 2:
-                    break
-
-                # sleep in the loop executor, this forces execution back into
-                # the event loop to avoid the block thread from starving the
-                # async loop
-                run_coroutine_threadsafe(sleep_wait(), self.loop).result()
+    @asyncio.coroutine
+    def async_block_till_done(self):
+        """Block till all pending work is done."""
+        count = 0
+        while True:
+            # Wait for the pending tasks are down
+            if len(self._pending_tasks) > 0:
+                yield from asyncio.gather(*self._pending_tasks, loop=self.loop)
 
-            complete.set()
+            # Verify the loop is empty
+            ret = yield from self.loop.run_in_executor(None, self._loop_empty)
```

Comment:
I run that inside event loop and self._loop_empty will never return true... a deadlock